### PR TITLE
feat: download-link AdSense popup with literal HTML markup

### PR DESCRIPTION
**Pull request by nahcnuj**
## Summary

Markdown 内でリンクテキストに「ダウンロード」を含み、href がダウンロード可能なファイル拡張子を指すリンクを、`<button popovertarget>` に置き換えます。クリック時に Popover API で広告付きポップアップを開きつつ、クライアントスクリプトが一時的な `<a>` でダウンロードを開始します。

著者は `[ダウンロード](./file.pdf)` と書くだけでよく、`downloadAdPopup: true` は remark プラグインが自動付与します。

### Markdown パイプライン（`app/lib/downloadLinkPlugin.ts`）

- **remark**: 対象リンクがある MDX に `downloadAdPopup: true` を付与（複数リンクでも 1 回）
- **rehype**: 該当 `<a>` を `<button popovertarget="download-ad-popup">` に変換
  - `data-download-href` に元の href を保持
  - 同一オリジンは `data-download`、クロスオリジンは `data-download-new-tab`
  - `.html` / `.mdx` などページ系拡張子は対象外

### UI・クライアント（`app/components/DownloadAdDialog.tsx`）

- Hono JSX でポップアップを SSR（AdSense スロット `2208436912`）
- ポップアップの開閉は `popovertarget` / `popovertargetaction` に任せる（スクリプトでは開かない）
- `setupDownloadAdPopup`（`app/client.ts` から呼び出し）
  - `document` へのイベント委譲で複数ボタンに対応
  - `startDownload`: 一時 `<a>.click()` でダウンロード開始
  - `prepareDownloadAdPopup`: ダイアログ内フォールバック `<a>` の href を更新（手動リトライ用）
- 1 ページに複数のダウンロードリンクがあっても、ポップアップは 1 つを共有

### レイアウト

- `articleMdxRenderer`: `downloadAdPopup === true` の記事に `<DownloadAdDialog />` を挿入
- `RootLayout`: 本番、または開発時の対象記事で AdSense ローダーを出力

## テスト

| 種別 | ファイル | 件数 |
|------|----------|------|
| 単体 | `app/lib/downloadLinkPlugin.test.ts` | 20 |
| 単体 | `app/components/DownloadAdDialog.test.ts` | 10 |
| E2E | `tests/e2e/download-link.test.ts` | 9 |
| VRT | `tests/e2e/download-ad-dialog.vrt.test.ts` | 3 |

- ボタンクリックによる実ダウンロード / クロスオリジンの新規タブは E2E で検証
- ポップアップ内フォールバック `<a>` のクリック挙動はブラウザの責務のためテストしない
- 必須マークアップのアサーションは各テストファイルに直接記述

### ビジュアルリグレッション（VRT）

`tests/e2e/download-ad-dialog.vrt.test.ts` でポップアップ表示時の見た目を 375px / 1280px / 1440px の 3 ビューポートでスナップショット比較します。

- AdSense 領域（`.download-ad-container`）はマスク
- `adsbygoogle.js` はブロック
- Linux（CI）と win32 のベースライン PNG をリポジトリに含める
- `npm run test:e2e`（Vitest）と `npm run test:vrt`（Playwright）は別スクリプト
- CI では `e2e` ジョブと `vrt` ジョブでそれぞれ実行

## フィクスチャ（`app/fixtures/`）

| ファイル | 用途 |
|----------|------|
| `essays/download-link.mdx` + `test.pdf` | 同一オリジン・単一リンク |
| `essays/download-links.mdx` | 同一ファイルへの複数リンク |
| `works/download-link-test.mdx` | クロスオリジン |

開発時は `devFixturesPlugin` が fixtures を `app/routes/` にコピーし、PDF はミドルウェアから配信します（`public/` には置かない）。

### CODE REVIEW

File: app/lib/downloadAdPopup.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+/**
+ * Shows the pre-rendered AdSense popup when Markdown download links are clicked.
+ * The native download behaviour of the link is not blocked.
+ */
+export function setupDownloadAdPopup({
+  whenReady,
+  getDownloadLinks,
+  getPopupDialog,
+  showPopup = (dialog) => {
+    dialog.showModal()
```
Comment (nahcnuj) [inline app/lib/downloadAdPopup.ts]: **[bug]** `showPopup` always calls `dialog.showModal()`. If the dialog is already open (user clicks a second download link on the same page, or re-clicks before closing), the HTML spec requires `showModal()` to throw `InvalidStateError`. That can surface as an uncaught exception in the click handler after `prepareDownloadAdPopup` has already run.

**Suggestion:** Guard with `if (!dialog.open) dialog.showModal()` (or close then reopen if you want a fresh modal), and add a unit test for the already-open case.

File: app/lib/rehypeDownloadLinks.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ */
+export function rehypeDownloadLinks() {
+  return (tree: Root) => {
+    visit(tree, 'element', (node) => {
+      if (!isDownloadLinkText(node)) return
+
+      node.properties = node.properties ?? {}
+      node.properties.download = ''
+      node.properties[DOWNLOAD_AD_DATA_ATTR] = ''
+      delete node.properties.target
```
Comment (nahcnuj) [inline app/lib/rehypeDownloadLinks.ts]: **[bug]** Download links unconditionally `delete node.properties.target`. For cross-origin URLs (e.g. the `works/download-link-test.mdx` fixture pointing at `https://example.com/sample.pdf`), browsers ignore `download=""` and follow `href` in the same tab. The user navigates away while the popup may flash, and no file download occurs—contrary to the "download started" copy and fallback-link design.

**Suggestion:** Only strip `target` (and set `download`) for same-origin/same-site URLs; keep `target="_blank"` (or another safe strategy) for external origins, or document and enforce that download links must be same-origin relative paths.

File: app/components/RootLayout.tsx [hunk 3]
```diff
@@ -104,7 +169,7 @@ const Layout = (props: PropsWithChildren<Meta>) => html`
   ${<Style>{rootStyle}</Style>}
   <link rel="alternate" type="application/rss+xml" title="www.nahcnuj.work" href="/feed.xml">
   ${import.meta.env.PROD && gtagSnippets.head}\
-  ${import.meta.env.PROD && adsenseSnippet}\
+  ${adsenseSnippet}\
```
Comment (nahcnuj) [inline app/components/RootLayout.tsx]: **[suggestion]** The AdSense loader script is now injected on every page in all environments (`${adsenseSnippet}`), whereas it was previously gated behind `import.meta.env.PROD`. Local dev and preview builds will load the real `adsbygoogle.js` client on non-download pages too.

**Suggestion:** Restore a PROD guard, or scope loading to pages with `downloadAdPopup` / an explicit E2E flag so dev traffic does not hit production ad tags globally.

File: app/lib/rehypeDownloadLinks.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import { visit } from 'unist-util-visit'
+export const DOWNLOAD_LINK_MARKER = 'ダウンロード'
+export const DOWNLOAD_AD_DATA_ATTR = 'data-download-ad'
+
+/** Returns true when the element is an anchor whose text contains 「ダウンロード」. */
+export function isDownloadLinkText(element: Element): boolean {
+  if (element.tagName !== 'a') return false
+  const href = element.properties?.href
+  if (typeof href !== 'string' || href.length === 0) return false
+  return hastToString(element).includes(DOWNLOAD_LINK_MARKER)
```
Comment (nahcnuj) [inline app/lib/rehypeDownloadLinks.ts]: **[suggestion]** Any anchor whose visible text merely contains the substring `ダウンロード` is treated as a download link (e.g. "ダウンロードページへ" linking to an HTML page). That page would get the popup dialog, `download=""`, and client listeners even when no file download is intended.

**Suggestion:** Tighten detection (e.g. require a file extension in `href`, an explicit frontmatter flag, or a more deliberate author convention) and add regression tests for negative cases.

File: app/components/DownloadAdDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import { DOWNLOAD_AD_FALLBACK_ID, DOWNLOAD_AD_POPUP_ID } from '../lib/downloadAdMarkup'
+import {
+  DOWNLOAD_DIALOG_CLASS,
+  DOWNLOAD_DIALOG_LABEL,
+  DOWNLOAD_FALLBACK_LINK_TEXT,
+} from '../lib/downloadAdPopup'
+import { ADSENSE_CLIENT_ID, DOWNLOAD_AD_SLOT } from '../lib/site'
+
+/** Pre-rendered download popup dialog with literal AdSense markup. */
+export default function DownloadAdDialog() {
```
Comment (nahcnuj) [inline app/components/DownloadAdDialog.tsx:139]: **[suggestion]** The full dialog + AdSense markup is duplicated in `DownloadAdDialog.tsx` (runtime) and `downloadAdMarkup.ts` (`downloadAdDialogMarkup`, used only in tests). The component tests assert the string module, not the component output, so JSX/template drift between the two paths would not be caught.

**Suggestion:** Have `DownloadAdDialog` render from a single shared markup builder (or test the component's HTML output directly).

File: vite.config.ts [hunk 6]
```diff
@@ -19,25 +21,38 @@ function devFixturesPlugin(): Plugin {
...
           const destPath = join(destDir, entry.name)
           if (entry.isDirectory()) {
-            copyMdxFiles(srcPath, destPath)
-          } else if (entry.name.endsWith('.mdx')) {
-            if (!destDirCreated) {
-              mkdirSync(destDir, { recursive: true })
-              destDirCreated = true
-            }
+            copyFixtureFiles(srcPath, destPath)
+          } else if (!entry.name.startsWith('.')) {
```
Comment (nahcnuj) [inline vite.config.ts:29]: **[suggestion]** `copyFixtureFiles` now copies every non-dot fixture file (including `test.pdf`) into `app/routes/`, but `.gitignore` only ignores `app/routes/**/*.mdx`. If the dev server exits uncleanly, a copied PDF could remain under `app/routes/` and be at risk of accidental commit.

**Suggestion:** Extend gitignore to all dev-copied fixture artifacts (e.g. `app/routes/**/*.pdf`) or copy only `.mdx` and rely on the fixture middleware for binary assets.

File: app/lib/remarkDownloadAdPopup.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    visit(tree, 'link', (node) => {
+      if (toString(node).includes(DOWNLOAD_LINK_MARKER)) {
+        hasDownloadLink = true
+      }
+    })
+
+    if (!hasDownloadLink) return
+
+    const yamlNode = tree.children.find((node) => node.type === 'yaml')
+    if (!yamlNode || yamlNode.type !== 'yaml') return
```
Comment (nahcnuj) [inline app/lib/remarkDownloadAdPopup.ts]: **[suggestion]** When a page has a 「ダウンロード」 link but no YAML frontmatter node, the remark plugin silently skips and `articleMdxRenderer` will not render `<DownloadAdDialog />`, while `rehypeDownloadLinks` still marks the anchor and `setupDownloadAdPopup` bails because the dialog is missing—popup never appears with no build-time signal.

**Suggestion:** Fail loudly in the remark plugin, or always render the dialog whenever `data-download-ad` links exist, so author mistakes are visible.

File: app/components/RootLayout.tsx [hunk 8]
```diff
@@ -52,6 +52,71 @@ const rootStyle = css`
...
+    display: flex;
+    justify-content: center;
+  }
+
+  .download-ad-dialog .download-ad-actions {
+    display: flex;
+    justify-content: flex-end;
+  }
+
+  .download-ad-dialog .download-ad-close-icon {
```
Comment (nahcnuj) [inline app/components/RootLayout.tsx:112]: **[nit]** `.download-ad-close-icon` uses `position: absolute` with `top`/`right`, but `.download-ad-dialog` has no `position: relative` (or similar), so the close button may anchor to an unexpected ancestor depending on layout.

**Suggestion:** Add `position: relative` to `.download-ad-dialog` (or the `<form>`) so the icon stays in the dialog corner.

File: .gitignore [hunk 9]
```diff
@@ -20,3 +20,7 @@ wrangler.toml
 yarn.lock
 pnpm-lock.yaml
 bun.lockb
+
+# Grok Build
```
Comment (nahcnuj) [inline .gitignore]: **[nit]** `# Grok Build`, `terminals/`, and `/*.html` are unrelated to the download-link / AdSense feature and broaden repo ignore rules without explanation in the PR.

**Suggestion:** Drop unrelated hunks from this PR or split them into a separate commit with rationale.

File: vite.config.ts [hunk 10]
```diff
@@ -19,25 +21,38 @@ function devFixturesPlugin(): Plugin {
...
+          return
+        }
+
+        const fixturePath = join(fixturesDir, relativePath)
+        if (!existsSync(fixturePath) || !statSync(fixturePath).isFile() || extname(fixturePath) === '.mdx') {
+          next()
+          return
+        }
+
+        createReadStream(fixturePath).pipe(res)
```
Comment (nahcnuj) [inline vite.config.ts]: **[nit]** The fixture static middleware pipes `createReadStream(fixturePath)` without setting `Content-Type`/`Content-Disposition` or handling stream `error` events; failures could hang or confuse clients for non-MDX fixture assets.

**Suggestion:** Set MIME type from `extname`, optionally `Content-Disposition: attachment` for PDFs used in download tests, and forward stream errors to `next(err)`.

File: app/lib/downloadAdPopup.ts (hunk 1 above)
Comment (nahcnuj) [reply app/lib/downloadAdPopup.ts]: 対応しました。showPopup のデフォルト実装に if (!dialog.open) ガードを追加し、ダイアログが既に開いている場合は showModal() を呼ばないようにしました。あわせてユニットテストを追加しています（f5cf47a）。

File: app/lib/rehypeDownloadLinks.ts (hunk 2 above)
Comment (nahcnuj) [reply app/lib/rehypeDownloadLinks.ts]: 対応しました。isSameOriginDownloadHref を追加し、同一オリジン（相対パス含む）のリンクのみ download 属性の付与と 	arget の削除を行うようにしました。クロスオリジンのリンクは 	arget=_blank を維持し、ポップアップ用の data-download-ad のみ付与します。テストも更新済みです。

File: app/components/RootLayout.tsx (hunk 3 above)
Comment (nahcnuj) [reply app/components/RootLayout.tsx]: 対応しました。本番では従来どおり全ページに AdSense ローダーを出力し、開発環境では downloadAdPopup: true のページ（ダウンロードリンクを含む fixture 等）にのみ限定するようにしました。_renderer から frontmatter.downloadAdPopup を RootLayout に渡しています。

File: app/lib/rehypeDownloadLinks.ts (hunk 4 above)
Comment (nahcnuj) [reply app/lib/rehypeDownloadLinks.ts]: 対応しました。リンクテキストに「ダウンロード」が含まれることに加え、href にファイル拡張子がある場合のみダウンロードリンクと判定する isDownloadLink ヘルパーを追加しました。remark / rehype の両プラグインで共有し、否定ケースのテストも追加しています。

File: app/components/DownloadAdDialog.tsx (hunk 5 above)
Comment (nahcnuj) [reply app/components/DownloadAdDialog.tsx:139]: 対応しました。DownloadAdDialog は downloadAdDialogMarkup() を html + raw() で描画するだけにし、ランタイムとテストで同じマークアップ文字列を共有するようにしました。

File: .gitignore (hunk 9 above)
Comment (nahcnuj) [reply .gitignore]: 対応しました。本 PR と無関係な Grok Build / terminals / *.html の .gitignore エントリを削除しました（a943779）。

File: app/lib/remarkDownloadAdPopup.ts (hunk 7 above)
Comment (nahcnuj) [reply app/lib/remarkDownloadAdPopup.ts]: 対応しました。ダウンロードリンクがあるのに YAML frontmatter がない場合、remark プラグインがビルド時にエラーを throw するようにしました。テストも追加済みです（f18ee96）。

File: app/components/RootLayout.tsx (hunk 8 above)
Comment (nahcnuj) [reply app/components/RootLayout.tsx:112]: 対応しました。.download-ad-dialog に position: relative を追加し、閉じるボタンがダイアログ内の右上に固定されるようにしました（156e5c3）。

File: vite.config.ts (hunk 10 above)
Comment (nahcnuj) [reply vite.config.ts]: 対応しました。fixture ミドルウェアで Content-Type（PDF は application/pdf）を設定し、ストリームの error イベントを next(err) に転送するようにしました（8b42737）。

File: vite.config.ts (hunk 6 above)
Comment (nahcnuj) [reply vite.config.ts:29]: 対応しました。dev 時に app/fixtures からコピーされる PDF を誤コミットしないよう、app/routes/**/*.pdf を .gitignore に追加しました（e41a017）。
